Bluetooth: Mesh: Configurable RPL module

Andrewpini opened the pull request:
Introduce configuration options for chosen RPL implementations.

This will allow introducing alternate persistent storage schemes
for the replay protection list.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

jhedberg commented on subsys/bluetooth/mesh/rpl.c:
```diff
@@ -164,183 +113,30 @@ void bt_mesh_rpl_clear(void)
 	 */
-	for (i = 0; i < ARRAY_SIZE(replay_list); i++) {
-		struct bt_mesh_rpl *rpl = &replay_list[i];
+	for (int i = 0; i < CONFIG_BT_MESH_CRPL; i++) {
+		rpl = bt_mesh_rpl_strg_iter(rpl);
```
If you are going to encapsulate RPL iteration, then it should be done properly, so that the caller doesn't need to know details like `CONFIG_BT_MESH_CRPL`, i.e. this could instead look something like:
```c
for (struct bt_mesh_rpl *rpl = bt_mesh_strg_iter(NULL); rpl != NULL; rpl = bt_mesh_strg_iter(rpl)) {
	...;
}
```
That said, I think you might get a bit more efficient code (and also be more consistent with the iterators we have elsewhere within the Zephyr Bluetooth stack) by introducing a foreach function with a callback pointer:

```c
/* Callback can return false to abort iteration, or true to continue iterating */
void bt_mesh_rpl_foreach(bool (*func)(struct bt_mesh_rpl *rpl, void *user_data), void *user_data);
```

Related - it doesn't seem right to me that the storage backend would also own the runtime storage of the RPL, i.e. the `replay_list` array. Or is your intention to also be able to replace the runtime management of the RPL with another implementation?

If you're only intending to encapsulate the persistent storage, it would seem more appropriate for `rpl.c` to retain ownership of the runtime storage. If you need to do iteration in the storage backend, then the backend needs to call the iterator function (which would be implemented in `rpl.c`).

If you're intending to encapsulate both runtime and persistent storage, it looks like the purpose of `rpl.c` becomes really small (just a bit over 100 lines of code), and I wonder if it'd make more sense to just push all responsibility to the storage backend.

Note: I'm just thinking out loud here - don't take these as some final commandments, but rather let me know your thoughts!

PavelVPV replied:
> Related - it doesn't seem right to me that the storage backend would also own the runtime storage of the RPL, i.e. the replay_list array. Or is your intention to also be able to replace the runtime management of the RPL with another implementation?

Yes, this is the idea. A storage mechanism that we are planning to use in nRF Connect SDK for RPL will replace both, the runtime and persistence management. At the same time we want to keep the internal API implementation unchanged as much as possible to avoid potential bugs. Do you think this implementation will impact on performance?

jhedberg replied:
In that case it feels a bit like we might not need the additional layer of abstraction, rather the existing `rpl.c` could be replaced with one of several possible `rpl_<backend>.c` implementations. If there's substantial amounts of code that multiple backends would need to duplicate, we could think of some alternate solutions to that, e.g. inline functions in `rpl.h` or similar.

Andrewpini replied:
> Related - it doesn't seem right to me that the storage backend would also own the runtime storage of the RPL, i.e. the replay_list array. Or is your intention to also be able to replace the runtime management of the RPL with another implementation?

> If you're only intending to encapsulate the persistent storage, it would seem more appropriate for rpl.c to retain ownership of the runtime storage. If you need to do iteration in the storage backend, then the backend needs to call the iterator function (which would be implemented in rpl.c).

> If you're intending to encapsulate both runtime and persistent storage, it looks like the purpose of rpl.c becomes really small (just a bit over 100 lines of code), and I wonder if it'd make more sense to just push all responsibility to the storage backend.

We have been working on an alternate persistent storage implementation in our Downstream. The storage mechanism for this module will write the entire RPL to flash in a single operation at shutdown, meaning that the module needs access to the memory object.

We have been back and forth on this in an internal review, and here are the main points:

If `rpl.c` owns the memory object for the RPL, the storage module would either need API access the memobj, or duplicate the entire RPL before storing it. To provide direct access to the memobj has previously been strongly discouraged since it will provide direct access to senitive data ([Link](https://github.com/zephyrproject-rtos/zephyr/pull/43441#discussion_r820671571)). Duplicating the list is neither a good option since it will double the amount of memory needed to ensure that the data is stored, especially when the RPL is large.

However, if the storage module owns the memobj then all these issues should disapear.

Andrewpini replied:
> In that case it feels a bit like we might not need the additional layer of abstraction, rather the existing `rpl.c` could be replaced with one of several possible `rpl_<backend>.c` implementations. If there's substantial amounts of code that multiple backends would need to duplicate, we could think of some alternate solutions to that, e.g. inline functions in `rpl.h` or similar.

Ok, on first thought I'm not necessarily opposed to this. It would remove the need to provide iteration API over the RPL array all together.

PavelVPV replied:
How about adding few functions to replace `bt_mesh_rpl_strg_iter`:
- _find first entry with such address or first empty entry_ for `bt_mesh_rpl_check`;
- _get next non-empty entry or return NULL_ for `bt_mesh_rpl_reset`;

jhedberg replied:
@PavelVPV that doesn’t sound like a bad approach. Would be good to see the code first however.

PavelVPV commented on subsys/bluetooth/mesh/rpl_storage_settings.c:
```diff
@@ -0,0 +1,279 @@
+
+struct bt_mesh_rpl *bt_mesh_rpl_next(struct bt_mesh_rpl *curr)
+{
+	int i = curr ? rpl_idx(curr) + 1 : 0;
+
+	if ((i < 0) || (i >= ARRAY_SIZE(replay_list) - 1)) {
+		BT_ERR("Invalid index for input RPL entry");
+		return NULL;
+	}
+
+	for (; i < ARRAY_SIZE(replay_list); i++) {
+		if (replay_list[i].src) {
+			return &replay_list[i];
+		}
+	}
+
+	return NULL;
```
```suggestion
    struct bt_mesh_rpl *rpl = curr ? curr + 1 : &replay_list[0];
    return rpl < &replay_list[CONFIG_BT_MESH_CRPL] ? rpl : NULL;
```

PavelVPV replied:
```suggestion
	for (int i = curr ? rpl_idx(curr) + 1 : 0; i < ARRAY_SIZE(replay_list); i++) {
		if (replay_list[i].src) {
			return &replay_list[i];
		}
	}

	return NULL;
```
Please disregard the first suggestion. I forgot about address check. Simply there is no need in if-statement if I see correctly.

Balaklaka commented on subsys/bluetooth/mesh/rpl_storage_settings.c:
```diff
@@ -0,0 +1,279 @@
+int bt_mesh_rpl_get(uint16_t addr, struct bt_mesh_rpl **rpl)
+{
+	for (int i = 0; i < ARRAY_SIZE(replay_list); i++) {
+		if (!replay_list[i].src || replay_list[i].src == addr) {
```
Will this return first free element regardless if there is an element with the addr later in the list. Could happend if clear is run.

Andrewpini replied:
Can't be 100% sure, but I think this duplicates 1:1 the previous implementation for `bt_mesh_rpl_check`. The previous implementation iterated over the RPL until it found either an empty entry slot or an existing entry. As soon as either one of those criteria was fulfilled it returned. https://github.com/zephyrproject-rtos/zephyr/blob/735ddeb4a8e26c6ce07d6aa5a0b77fffc559b2a8/subsys/bluetooth/mesh/rpl.c#L124-L157

alxelax replied:
It replicates but it seems like a bug. Some functionality requests existing RPL entry but gets an empty cell. Is this a potential attack vector for mesh devices? Especially after the reset function that cleans up some of the RPL cells.

PavelVPV replied:
Right, then this is a bug in the initial implementation as well. 

Shouldn't this:

> Even more, since RPL and RPL storage are split the assumption about no empty entries in between is a kind of hidden knowledge.

be responsibility of the RPL storage?

alxelax replied:
This is possible to solve in both:
1. Implement consideration of having empty entries in between in RPL
2. Implement some `defrag` in RPL storage to keep list without empty cells.

and probably as a separate commit or even PR.

PavelVPV replied:
I'll take care of this bug.

Andrewpini commented on subsys/bluetooth/mesh/rpl.c:
```diff
@@ -121,226 +64,70 @@ bool bt_mesh_rpl_check(struct bt_mesh_net_rx *rx,
+	}
 
-			return false;
+	if (!rpl) {
```
I see your point. I was considering this check back and forth for that exact reason. It really should not happen that `rpl = NULL` here, but I thought it might be wise since this is a pointer and there might be multiple implementations for the body of the `bt_mesh_rpl_get`.

PavelVPV replied:
If I don't miss anything, `bt_mesh_rpl_get` could just return pointer to `struct bt_mesh_rpl_get` instead of int.

PavelVPV reviewed (commented):
Follow up on internal discussion with @Andrewpini:
- Move implementation of `bt_mesh_rpl_reset` and `bt_mesh_rpl_clear` into `rpl_storage_settings.c`;
- Remove `bt_mesh_rpl_next`;
- Leave only `bt_mesh_rpl_strg_store` and `bt_mesh_rpl_get` in `rpl_storage.h`.

alxelax commented on subsys/bluetooth/mesh/rpl_storage_settings.c:
```diff
@@ -0,0 +1,293 @@
+{
+	int i;
+
+	for (i = 0; i < CONFIG_BT_MESH_CRPL; i++) {
```
```suggestion
	for (int i = 0; i < CONFIG_BT_MESH_CRPL; i++) {
```

alxelax commented on subsys/bluetooth/mesh/rpl_storage_settings.c:
```diff
@@ -0,0 +1,293 @@
+{
+	int i;
+
+	for (i = 0; i < CONFIG_BT_MESH_CRPL; i++) {
```
```suggestion
	for (int i = 0; i < CONFIG_BT_MESH_CRPL; i++) {
```

alxelax commented on subsys/bluetooth/mesh/rpl_storage_settings.c:
```diff
@@ -0,0 +1,293 @@
+		bt_mesh_settings_store_cancel(BT_MESH_SETTINGS_RPL_PENDING);
+	}
+
+	for (i = 0; i < CONFIG_BT_MESH_CRPL; i++) {
```
```suggestion
	for (int i = 0; i < CONFIG_BT_MESH_CRPL; i++) {
```

alxelax commented:
Seems no clear boundary between rpl and rpl storage functionality. Part of rpl logic and API is in storage implementation.

jhedberg commented on subsys/bluetooth/mesh/rpl_storage.h:
```diff
@@ -0,0 +1,22 @@
+ * @param[in] rpl RPL entry to store.
+ * @param[in] force Flag for forcing the storage.
+ */
+void bt_mesh_rpl_strg_store(struct bt_mesh_rpl *rpl, bool force);
```
Could you explain the usage of the `force` parameter a bit more, since you don't seem to be using it in any way in the implementation?

Another thing I'm not quite happy about is that you use `storage` in the filename and `STORAGE` in Kconfig, but `strg` in the API. I'm not a big fan of long symbol names, but would it be that bad to keep this consistent and spell out the whole word here as well?

Andrewpini replied:
And now I see there was a bug in `bt_mesh_rpl_storage_store` aswell

alxelax commented on subsys/bluetooth/mesh/rpl_storage_settings.c:
```diff
@@ -0,0 +1,293 @@
+
+	err = bt_mesh_settings_set(read_cb, cb_arg, &rpl, sizeof(rpl));
+	if (err) {
+		BT_ERR("Failed to set `net`");
```
I'd suggest "Failed to set `rpl`"

jhedberg commented:
> > Seems no clear boundary between rpl and rpl storage functionality. Part of rpl logic and API is in storage implementation.
> 
> I'm starting to lean that way myself. Maybe it would be better to just skip the storage abbstraction level, use separate RPL modules and a common header in `rpl.h`

The way I see the current proposal is that it's not even trying to create a clean boundary for RPL functionality. The storage-specific module is the primary implementation of the RPL and what remains in `rpl.c` is simply an attempt to avoid duplicating code in multiple storage-specific modules.

That said, I already suggested earlier removing this additional layer and having just a high level choice of selecting the RPL implementation (including storage). Looking at the latest version of this PR there's only some 100 lines remaining in `rpl.c`, so I think there's a pretty good case for eliminating the extra abstraction layer.

Andrewpini commented:
I've added a significantly less complex solution where it is possible to introduce an alternate version of rpl.c. What do you think @jhedberg?

Andrewpini committed:
Bluetooth: Mesh: Configurable RPL module

Introduce configuration options for chosen RPL implementations.

This will allow introducing alternate persistent storage schemes
for the replay protection list.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

LingaoM commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -289,6 +289,46 @@ config BT_MESH_CRPL
+
+endchoice
+
+if BT_MESH_RPL_STORAGE_MODE_SETTINGS && BT_SETTINGS
```
```suggestion
if BT_MESH_RPL_STORAGE_MODE_SETTINGS
```

LingaoM commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -289,6 +289,46 @@ config BT_MESH_CRPL
+	  will cause the device to not perform the replay protection
+	  required by the spec.
+
+endif # BT_MESH_RPL_STORAGE_MODE_SETTINGS && BT_SETTINGS
```
```suggestion
endif # BT_MESH_RPL_STORAGE_MODE_SETTINGS
```

Andrewpini commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -289,6 +289,46 @@ config BT_MESH_CRPL
+
+config BT_MESH_RPL_STORAGE_MODE_SETTINGS
+	bool "Persistent storage of RPL in Settings"
+	help
```
It might seem a little akward to not add this dependency, but the idea is that this is **not** dependent on BT_SETTINGS. Like described in the help text for this option, if `BT_SETTINGS != y` rpl.c will provide a non-persistent alternative for RPL (Just like the previous implementation).

Andrewpini replied:
It could be solved by something like this:
```
choice BT_MESH_RPL_STORAGE_MODE
	prompt "Replay protection list storage mode"
	default BT_MESH_RPL_STORAGE_MODE_SETTINGS

config BT_MESH_RPL_STORAGE_MODE_SETTINGS
	bool "Persistent storage of RPL in Settings"
	depends on BT_SETTINGS
	help
	  Persistent storage of RPL in Settings. 

config BT_MESH_RPL_STORAGE_MODE_NON_PERSIST
	bool "Non-persistent storage of RPL "
	depends on !BT_SETTINGS
	help
	  Storage of RPL in volatile memory only.

endchoice
```

```
if(CONFIG_BT_MESH_RPL_STORAGE_MODE_SETTINGS OR CONFIG_BT_MESH_RPL_STORAGE_MODE_NON_PERSIST)
  zephyr_library_sources(rpl.c)
endif()
```
But I think that the added bloat and complexity might not justify it.

jhedberg commented on subsys/bluetooth/mesh/settings.c:
```diff
@@ -111,12 +117,12 @@ void bt_mesh_settings_store_schedule(enum bt_mesh_settings_flag flag)
+		timeout_ms = RPL_STORE_TIMEOUT * MSEC_PER_SEC;
 	} else {
-		timeout_ms = CONFIG_BT_MESH_STORE_TIMEOUT * MSEC_PER_SEC;
+		timeout_ms = RPL_STORE_TIMEOUT * MSEC_PER_SEC;
```
Doesn't this potentially expand to `-1 * MSEC_PER_SEC`?

Andrewpini replied:
You are right, it is a bug. Should be `BT_MESH_STORE_TIMEOUT`
